src/compiler/scanner.ts

```diff
@@ -1458,6 +1467,70 @@ module ts {
             return token = scanTemplateAndSetTokenValue();
         }
 
+        function reScanLessThanSlashToken(): SyntaxKind {
+            pos = tokenPos;
+            return token = scanJsxToken();
+        }
+        
+        function reScanJsxToken(): SyntaxKind {
+            pos = tokenPos = startPos;
+            return token = scanJsxToken();
+        }
+
+        function scanJsxToken(): SyntaxKind {
+            startPos = tokenPos = pos;
+
+            if (pos >= end) {
+                return token = SyntaxKind.EndOfFileToken;
+            }
+
+            let char = text.charCodeAt(pos);
+            if (char === CharacterCodes.lessThan) {
+                if (text.charCodeAt(pos + 1) === CharacterCodes.slash) {
+                    pos += 2;
+                    return token = SyntaxKind.LessThanSlashToken;
+                }
+                else {
+                    pos++;
+                    return token = SyntaxKind.LessThanToken;
+                }
+            }
+            else if (char === CharacterCodes.openBrace) {
+                pos++;
+                return token = SyntaxKind.OpenBraceToken;
+            }
+            else {
+                while (pos < end) {
+                    pos++;
+                    char = text.charCodeAt(pos);
+                    if ((char === CharacterCodes.openBrace) || (char === CharacterCodes.lessThan)) {
+                        break;
+                    }
+                }
+                return token = SyntaxKind.JsxText;
+            }
+        }
+
+        // Scans a JSX identifier; these differ from normal identifiers in that
+        // they allow dashes
+        function reScanJsxIdentifier(): SyntaxKind {
+            if (token === SyntaxKind.Identifier) {
+                let first = true;
+                while (pos < end) {
+                    let ch = text.charCodeAt(pos);
+                    if (ch === CharacterCodes.minus || (first ? isIdentifierStart(ch) : isIdentifierPart(ch))) {
+                        tokenValue += String.fromCharCode(ch);

```

Comment:
Ewwww.  expensive string concat.  Can we not do this afterwards as one substr call?